web/src/main/java/com/microsoft/applicationinsights/web/internal/correlation/TraceContextCorrelation.java

```diff
@@ -0,0 +1,288 @@
+                outGoingTraceParent = new Traceparent(0, incomingTraceparent.getTraceId(),
+                    null, incomingTraceparent.getTraceFlags());
+
+                // set id of this request
+                requestTelemetry.setId(outGoingTraceParent.getTraceId() + "-" + outGoingTraceParent.getSpanId());
+
+                // represents the trace-id of this distributed trace
+                requestTelemetry.getContext().getOperation().setId(outGoingTraceParent.getTraceId());
+
+                // represents the parent-id of this request which is combination of traceId and incoming spanId
+                requestTelemetry.getContext().getOperation().setParentId(outGoingTraceParent.getTraceId() + "-" +
+                    incomingTraceparent.getSpanId());
+
+            }
+
+            String tracestate = request.getHeader(CORRELATION_COMPANION_HEADER_NAME);
+            Tracestate tracestateObject = Tracestate.fromString(tracestate);
+            Map<String, String> tracestatePropertiesMap = getPropertiesMap(tracestateObject);
+
+            // create outbound tracestate to be propagated to downstream calls
+            Tracestate outboundTracestate = createOutboundTracestate(tracestatePropertiesMap, getAppId());
+            ThreadContext.getRequestTelemetryContext().setTracestate(outboundTracestate);
+
+            // Let the callee know the caller's AppId
+            addTracestateInResponseHeader(response);
+
+        } catch (Exception e) {
+            InternalLogger.INSTANCE.trace("unable to perform correlation :%s", ExceptionUtils.
+                getStackTrace(e));
+        }
+    }
+
+    /**
+     * This adds the tracestate in response header so that the Callee can know what is the caller's AppId.
+     */
+    private static void addTracestateInResponseHeader(HttpServletResponse response) {
+
+        if (response.containsHeader(CORRELATION_COMPANION_HEADER_NAME)) {
+            return;
+        }
+
+        String appId = getAppId();
+        if (appId == null || appId.isEmpty()) {
+            return;
+        }
+
+        // TODO: should we propagate the entire tracestate here?
+        Tracestate tracestate = new Tracestate("az" + appId);
+
+        response.addHeader(CORRELATION_COMPANION_HEADER_NAME, tracestate.toString());
+    }
+
+    /**
+     * Retrieves the appId for the current active config's instrumentation key.
+     */
+    public static String getAppId() {
+
+        String instrumentationKey = TelemetryConfiguration.getActive().getInstrumentationKey();
+        String appId = InstrumentationKeyResolver.INSTANCE.resolveInstrumentationKey(instrumentationKey);
+
+        //it's possible the appId returned is null (e.g. async task is still pending or has failed). In this case, just
+        //return and let the next request resolve the ikey.
+        if (appId == null) {
+            InternalLogger.INSTANCE.trace("Application correlation Id could not be retrieved (e.g. task may be pending or failed)");
```

Comment:
On the fence: Should this be a warning?

Reply:
Trace is good enough here I think. ID should be resolved eventually. It's just that async task might take some time to complete.